src/trajectory.jl

```diff
@@ -573,13 +595,101 @@ using the generalised no-U-turn criterion.
 Ref: https://arxiv.org/abs/1701.02434
 """
 function isterminated(h::Hamiltonian, t::BinaryTree{<:GeneralisedNoUTurn})
-    # z0 is starting point and z1 is ending point
-    z0, z1 = t.zleft, t.zright
     rho = t.c.rho
-    s = (dot(rho, ∂H∂r(h, -z0.r)) >= 0) || (dot(-rho, ∂H∂r(h, z1.r)) >= 0)
+    s = generalised_uturn_criterion(rho, ∂H∂r(h, t.zleft.r), ∂H∂r(h, t.zright.r))
     return Termination(s, false)
 end
 
+"""
```

Comment:
The function signature in the comment is different from the actual code. In fact, I think we can simplify all to something like:

```julia
function isterminated(
    h::Hamiltonian, t::T, tleft::T, tright::T
) where {T<:BinaryTree{<:StrictGeneralisedNoUTurn}}
    ...
end
```

which make the code a bit neater.

Reply:
This also applies to `check_left_subtree ` and `check_right_subtree`.